Respect viewport position in coordinate conversion functions

Novakasa opened the pull request:
# Objective

- In `Camera::viewport_to_world_2d`, `Camera::viewport_to_world`, `Camera::world_to_viewport` and `Camera::world_to_viewport_with_depth`, the results were incorrect when the `Camera::viewport` field was configured with a viewport position that was non-zero. This PR attempts to correct that.
- Fixes #16200 

## Solution

- This PR now takes the viewport position into account in the functions mentioned above.
- Extended `2d_viewport_to_world` example to test the functions with a dynamic viewport position and size, camera positions and zoom levels. It is probably worth discussing whether to change the example, add a new one or just completely skip touching the examples.

## Testing

Used the modified example to test the functions with dynamic camera transform as well as dynamic viewport size and position.

alice-i-cecile commented:
> ## Migration Guide
> Not sure if this does this need a migration guide entry, since this can probably be considered a bugfix. 
> 
> This PR changes the result of the functions `Camera::viewport_to_world_2d`, `Camera::viewport_to_world`, `Camera::world_to_viewport` and `Camera::world_to_viewport_with_depth` in the case when the viewport position is not `UVec2::ZERO`. With a viewport position of `UVec2::ZERO`, the behaviour is not changed.

Good migration guide, but yes, bug fix :)

mockersf commented on examples/2d/2d_viewport_to_world.rs:
```diff
@@ -1,12 +1,16 @@
 //! This example demonstrates how to use the `Camera::viewport_to_world_2d` method.
 
-use bevy::{color::palettes::basic::WHITE, prelude::*};
+use bevy::render::camera::Viewport;
+use bevy::{
+    color::palettes::{basic::WHITE, css::GREEN},
+    prelude::*,
+};
```
```suggestion
use bevy::{
    color::palettes::{basic::WHITE, css::GREEN},
    prelude::*,
    render::camera::Viewport,
};
```

tychedelia reviewed (changes requested):
Thanks for taking the time to update the example to make sure this doesn't regress. The example looks a little funny on macOS, likely due to window scale factor, although seems to be functional. I'd suggest testing with scale factor = 2.0 to emulate macOS.

![image](https://github.com/user-attachments/assets/1e3abbc2-35b1-4bcd-8711-ebcc098b5964)

Novakasa commented:
Polished the example quite a bit
- Initialize viewport relative to physical window size (should hopefully fix MacOS weirdness)
- Controls now in `FixedUpdate` and respect fixed delta time
- Simplified viewport bounds visualization with a background mesh
- Draw cursor after Transform propagation to get rid of camera control lag
- Add comments

Novakasa commented:
@tychedelia would appreciate another test of the example, since I think the scale_override behaviour one Windows might be bit different than MacOS defaults.

tychedelia reviewed (approved):
Looks great and even more improved! Thanks.

Novakasa commented:
Rebased & fixed merge conflict, switched from `get_single()` to `single()`

Technically, it can be argued that these functions should rather be called `window_to_world_` rather than `viewport_to_world_`, but I would be fine with either, not sure if it would be worth introducing a breaking change over.
